### PR TITLE
[FIX] account: calculation of amount residual

### PR DESCRIPTION
**Pull request by nim-odoo**
When an account move line is reconciled with debits and credits, both are
subtracted from the original amount. This is not logical, since if the
line is a credit (debit) we should subtract (add) the debits and add
(subtract) the credits.

opw-667137

### CODE REVIEW

**Comment by qdp-odoo** [comment]
hello, i tested and I found that the 1st solution is what we would like to go for... But the implementation does satisfy me: ideally, it should be done on the account.move.line object, at the reconciliation time and not after.

Indeed, you can take the following use case: 
* company currency = usd, with rate 1.5 on 01/01/2016 and 2.0 on 02/02/2016
* invoice of 1000€ made on 01/01/2016 (1500$)
* customer payment of 800€ on 02/02/2016 (1600$)
If you register the payment directly on the invoice, or 
* create it manually through sales\customer payments 
* then reconcile manually the invoice and the payment (go in the journal items list, select both the receivable lines of the invoice and payment, click 'reconcile' under the 'actions' button)
=> you should have exactly the same effect as if you register the payment directly with the register payment on the invoice

So in short: I agree with the solution 1, but it's not done at the right place I think.

**Comment by nim-odoo** [comment]
@qdp-odoo : thanks, I'll correct it.

**Comment by nim-odoo** [comment]
> it should be done on the account.move.line object, at the reconciliation time and not after.

I don't really see how this can work. The write-off is created at the very end of the process, based on the remaining moves: https://github.com/odoo/odoo/blob/9.0/addons/account/models/account_move.py#L808

When the accounting journal of the invoice has no alternate currency, the reconciliation is done in USD. However, in USD, the customer has paid more than necessary, so it means that the payment itself was enough to cover the invoice (at least in USD, but not in EUR). Therefore, at the end of the process, the only remaining move is the payment move, so the write-off is created based on the residual amount of the payment. At that point, you have absolutely no idea that you are supposed to link a write-off to the invoice as well.

When the accounting journal of the invoice has an alternate currency (EUR in this case), the result is different since the reconciliation is done in EUR. The write-off is then linked to the invoice, not the payment anymore. By the way, in this case, two exchange difference entries are created (one linked to the invoice, one linked to the write-off). See: http://dl.free.fr/rJaD3s20i

Basically, my questions are: 
- in which currency are we supposed to do the reconciliation? The company currency? The invoice currency? Both?
- when is an invoice considered as paid? When the residual amount in the invoice currency is zero? When the residual amount in the company currency is zero? When both are zero?

I guess both should be consistent: if you consider an invoice as paid when the residual amount in the invoice currency is zero (as it is the case now) but you perform the reconciliation in the company currency, inconsistencies such as this one will arise.

Other issues:
- What to do if the payment is done in a third currency, like CHF?
- What if the payment needs to be associated to more than one invoice?

Thanks!

**Comment by treviser** [comment]
@nim-odoo Just to answer one of your questions:

> when is an invoice considered as paid? When the residual amount in the invoice currency is zero? When the residual amount in the company currency is zero? When both are zero?

From an accounting and a legal point of view, an invoice is paid when the full amount in the invoice currency is paid (when the residual amount in the invoice currency is zero). If the residual amount in the company currency is not equal to zero, you have an exchange difference.

**Comment by nim-odoo** [comment]
@treviser : thanks for the info!

**Comment by treviser** [comment]
You are welcome! More answers:

> in which currency are we supposed to do the reconciliation? The company currency? The invoice currency? Both?
Again, the invoice currency is only relevant. The invoice is part of a contractual relationship between the seller and the buyer, the company currency is (only) a legal requirement towards authorities and a business need, but has no legal relationship with customers and/or vendors. If you invoice in a particular currency with the explicit or implicit mutual understanding that the receivable is payable in the denominated currency then this is the contract. Otherwise you would have to state something like "amount in currency X or the equivalent in any other currency based on the exchange rate as defined ...."

> What to do if the payment is done in a third currency, like CHF?
Then you should reconcile invoice currency amount against payment currency amount converted to invoice currency amount. How to achieve that is another question.

> What if the payment needs to be associated to more than one invoice?
The answer is clear now, but how to achieve it must be further elaborated.

**Comment by fpodoo** [comment]
>> when is an invoice considered as paid? When the residual amount in the
invoice currency is zero? When the residual amount in the company currency
is zero? When both are zero?
>
> From an accounting and a legal point of view, an invoice is paid when the
full amount in the invoice currency is paid (when the residual amount in
the invoice currency is zero). If the residual amount in the company
currency is not equal to zero, you have an exchange difference.

Nicolas, ask QDP as he fixed issues recently on reconciliation and exchange
difference.

In odoo, an invoice is paid when it's receivables entries are fully
reconciled. To fully reconcile, you need to have company currency AND
second currency net to 0. (Which sometimes requires an exchange difference
to balance both currencies)

If the company currency receivable entries are not fully reconciled (not
balance to 0), it's not paid (even if the invoice currency is cleared)

Note that I wrote "fully reconciled", not reconciled. It's important as
this can be tricky in complex scenarios (3 invoices, two payments)

**Comment by nim-odoo** [comment]
@fpodoo : yes we have done some extensive testing before his commit on several issues I had in the past.

BTW, thanks for the functional information

**Comment by treviser** [comment]
> In odoo, an invoice is paid when it's receivables entries are fully reconciled. To fully reconcile, you need to have company currency AND second currency net to 0. (Which sometimes requires an exchange difference to balance both currencies) 

I do not agree. Paid is paid and that's when the contractual relationship is fulfilled under the contractual conditions, means also the currency.
I do agree that the complete transaction in company currency must result in zero too, but that's achieved by creating/booking the necessary residuals in company currency at the time when full payment in the contractual currency is accomplished.

**Comment by treviser** [comment]
This might look like nitpicking, but it is especially also important for the determination of accounting dates of "automatic" postings, such as write-offs, exchange differences and taxes (see my findings here for example: https://github.com/odoo/odoo/issues/9972 )
